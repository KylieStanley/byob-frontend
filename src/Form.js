import React, { Component } from 'react'

export default class Form extends Component {
  constructor() {
    super()
    this.state = {
      method: 'GET',
      path: '/api/v1/states',
      databaseObj: {},
      idRequired: false
    }
  }

  submitRequest = (e) => {
    console.log('request')
    e.preventDefault()

    const url = `https://byob-us-music-festivals.herokuapp.com${this.state.path}`

    if (url.includes('states')) {
      this.setState({
        databaseObj: {
          state: 'Arizona',
          number_of_music_festivals: 10,
          major_airport: 'Phoenix Sky Harbor International Airport',
          tourism_website: 'https://www.visitarizona.com'
        }
      })
    } else if (url.includes('festivals')) {
      this.setState({
        databaseObj: { 
          festival_name: 'Arizona Roots',
          start_end_dates: '2/9/19-2/10/19', 
          city: 'Chandler', 
          image: 'https://1b7ta73fjmj23201tc3suvsi-wpengine.netdna-ssl.com/wp-content/uploads/2016/10/California-Roots-Hero-Pic-600x400.jpg' 
        }
      })
    }

    switch (this.state.method) {
      case 'GET':
        this.submitGet(url)
        break
      case 'POST':
        this.submitPost(url)
        break
      case 'PUT':
        this.submitPut(url)
        break
      case 'DELETE':
        this.submitDelete(url)
        break
    }
  }


  submitGet = async (url) => {
    const response = await fetch(url, {
      method: this.state.method,
    })
    const result = await response.json()
    this.props.handleResult(result)
  }


  submitPost = async (url) => {
    const response = await fetch(url, {
      method: this.state.method,
      headers: {
        'Content-Type': 'application/json'
      },
       body: JSON.stringify(this.state.databaseObj)
    })
    const result = await response.json()
    this.props.handleResult(result)
  }


  submitPut = async (url) => {
    const response = await fetch(url, {
      method: this.state.method,
      headers: {
        'Content-Type': 'application/json'
      },
       body: JSON.stringify(this.state.databaseObj)
    })
    const result = await response.json()
    this.props.handleResult(result)
  }


  submitDelete = async (url) => {
    const response = await fetch(url, {
      method: this.state.method
    })
    const result = await response.json()
    this.props.handleResult(result)
  }

  changeIdRequiredState = (event) => {
    event.preventDefault()

    event.currentTarget.childNodes.forEach(option => {

      if (option.selected) {
        this.setState({ path: option.innerText})
        if (option.className === 'option-id') {
          this.setState({ idRequired: true })
        } else {
          this.setState({ idRequired: false })
        }
      }
    })
  }

  updatePath = (e) => {
    e.preventDefault()

    const path = this.state.path.replace(':id', e.currentTarget.value)
    this.setState({ path })
  }

  changeInputStyleOnStateChange = () => {
    if (this.state.idRequired) {
      return <input onChange={ this.updatePath } placeholder="Enter an ID number" type='number'/>
    } else {
      return 
    }
  }

  render() {
    return (
      <form onSubmit={ this.submitRequest }>
        <select>
          <option>GET</option>
          <option>POST</option>
          <option>PUT</option>
          <option>DELETE</option>
        </select>
        <select onChange={this.changeIdRequiredState}>
          <option>/api/v1/states</option>
          <option className='option-id'>/api/v1/states/:id</option>
          <option>/api/v1/festivals</option>
          <option className='option-id'>/api/v1/states?state_id=:id</option>
          <option className='option-id'>/api/v1/states/:id/festivals</option>
          <option className='option-id'>/api/v1/festivals/:id</option>
        </select>
        { this.changeInputStyleOnStateChange() }
        <button type="submit">Send</button>
      </form>
    )
  }
}
