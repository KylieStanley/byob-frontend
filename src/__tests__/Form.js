import React from 'react'
import { shallow } from 'enzyme'
import Form from '../Form'


describe('Form', () => {
  let wrapper
  let mockEvent

  beforeEach(() => {
    wrapper = shallow(<Form handleResult={ jest.fn() } />)
    mockEvent = {
      preventDefault: jest.fn(),
      target: {
        options: [{value: "DELETE"}, {value: "POST"}],
        selectedIndex: 1,
        value: 2
      }
    }
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should update the method in state', () => {
    wrapper.instance().updateMethod(mockEvent)
    expect(wrapper.instance().state.method).toEqual('POST')
  })

  it('should submit a request on submit of the form', () => {
    const spy = jest.spyOn(wrapper.instance(), 'submitRequest')
    wrapper.instance().forceUpdate()
    wrapper.find('form').simulate('submit', {
      preventDefault: () => {}
    })

    expect(spy).toHaveBeenCalled()
  })

  it('should run change updateMethod on change of the select', () => {
     const spy = jest.spyOn(wrapper.instance(), 'updateMethod')
    wrapper.instance().forceUpdate()
    wrapper.find('select').at(0).simulate('change', mockEvent)
    expect(spy).toHaveBeenCalled()
  })

  it('should check the method and run the appropriate api call', () => {
    const spy = jest.spyOn(wrapper.instance(), 'submitGet')
    wrapper.instance().checkMethods()
    expect(spy).toHaveBeenCalled()
  })

  it('should update the path for endpoints requiring an id', () => {
    wrapper.instance().setState({ path: '/api/v1/states/:id'})
    wrapper.instance().updatePath(mockEvent)
    expect(wrapper.instance().state.path).toEqual('/api/v1/states/2')
  })
})
