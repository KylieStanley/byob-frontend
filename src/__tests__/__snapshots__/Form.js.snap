// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Form
    handleResult={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <select>
          <option>
            GET
          </option>
          <option>
            POST
          </option>
          <option>
            PUT
          </option>
          <option>
            DELETE
          </option>
        </select>,
        <select
          onChange={[Function]}
        >
          <option>
            /api/v1/states
          </option>
          <option
            className="option-id"
          >
            /api/v1/states/:id
          </option>
          <option>
            /api/v1/festivals
          </option>
          <option
            className="option-id"
          >
            /api/v1/states?state_id=:id
          </option>
          <option
            className="option-id"
          >
            /api/v1/states/:id/festivals
          </option>
          <option
            className="option-id"
          >
            /api/v1/festivals/:id
          </option>
        </select>,
        undefined,
        <button
          type="submit"
        >
          Send
        </button>,
      ],
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <option>
              GET
            </option>,
            <option>
              POST
            </option>,
            <option>
              PUT
            </option>,
            <option>
              DELETE
            </option>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "GET",
            },
            "ref": null,
            "rendered": "GET",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "POST",
            },
            "ref": null,
            "rendered": "POST",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "PUT",
            },
            "ref": null,
            "rendered": "PUT",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "DELETE",
            },
            "ref": null,
            "rendered": "DELETE",
            "type": "option",
          },
        ],
        "type": "select",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <option>
              /api/v1/states
            </option>,
            <option
              className="option-id"
            >
              /api/v1/states/:id
            </option>,
            <option>
              /api/v1/festivals
            </option>,
            <option
              className="option-id"
            >
              /api/v1/states?state_id=:id
            </option>,
            <option
              className="option-id"
            >
              /api/v1/states/:id/festivals
            </option>,
            <option
              className="option-id"
            >
              /api/v1/festivals/:id
            </option>,
          ],
          "onChange": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "/api/v1/states",
            },
            "ref": null,
            "rendered": "/api/v1/states",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "/api/v1/states/:id",
              "className": "option-id",
            },
            "ref": null,
            "rendered": "/api/v1/states/:id",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "/api/v1/festivals",
            },
            "ref": null,
            "rendered": "/api/v1/festivals",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "/api/v1/states?state_id=:id",
              "className": "option-id",
            },
            "ref": null,
            "rendered": "/api/v1/states?state_id=:id",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "/api/v1/states/:id/festivals",
              "className": "option-id",
            },
            "ref": null,
            "rendered": "/api/v1/states/:id/festivals",
            "type": "option",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "/api/v1/festivals/:id",
              "className": "option-id",
            },
            "ref": null,
            "rendered": "/api/v1/festivals/:id",
            "type": "option",
          },
        ],
        "type": "select",
      },
      undefined,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Send",
          "type": "submit",
        },
        "ref": null,
        "rendered": "Send",
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <select>
            <option>
              GET
            </option>
            <option>
              POST
            </option>
            <option>
              PUT
            </option>
            <option>
              DELETE
            </option>
          </select>,
          <select
            onChange={[Function]}
          >
            <option>
              /api/v1/states
            </option>
            <option
              className="option-id"
            >
              /api/v1/states/:id
            </option>
            <option>
              /api/v1/festivals
            </option>
            <option
              className="option-id"
            >
              /api/v1/states?state_id=:id
            </option>
            <option
              className="option-id"
            >
              /api/v1/states/:id/festivals
            </option>
            <option
              className="option-id"
            >
              /api/v1/festivals/:id
            </option>
          </select>,
          undefined,
          <button
            type="submit"
          >
            Send
          </button>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <option>
                GET
              </option>,
              <option>
                POST
              </option>,
              <option>
                PUT
              </option>,
              <option>
                DELETE
              </option>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "GET",
              },
              "ref": null,
              "rendered": "GET",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "POST",
              },
              "ref": null,
              "rendered": "POST",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "PUT",
              },
              "ref": null,
              "rendered": "PUT",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "DELETE",
              },
              "ref": null,
              "rendered": "DELETE",
              "type": "option",
            },
          ],
          "type": "select",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <option>
                /api/v1/states
              </option>,
              <option
                className="option-id"
              >
                /api/v1/states/:id
              </option>,
              <option>
                /api/v1/festivals
              </option>,
              <option
                className="option-id"
              >
                /api/v1/states?state_id=:id
              </option>,
              <option
                className="option-id"
              >
                /api/v1/states/:id/festivals
              </option>,
              <option
                className="option-id"
              >
                /api/v1/festivals/:id
              </option>,
            ],
            "onChange": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "/api/v1/states",
              },
              "ref": null,
              "rendered": "/api/v1/states",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "/api/v1/states/:id",
                "className": "option-id",
              },
              "ref": null,
              "rendered": "/api/v1/states/:id",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "/api/v1/festivals",
              },
              "ref": null,
              "rendered": "/api/v1/festivals",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "/api/v1/states?state_id=:id",
                "className": "option-id",
              },
              "ref": null,
              "rendered": "/api/v1/states?state_id=:id",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "/api/v1/states/:id/festivals",
                "className": "option-id",
              },
              "ref": null,
              "rendered": "/api/v1/states/:id/festivals",
              "type": "option",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "/api/v1/festivals/:id",
                "className": "option-id",
              },
              "ref": null,
              "rendered": "/api/v1/festivals/:id",
              "type": "option",
            },
          ],
          "type": "select",
        },
        undefined,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Send",
            "type": "submit",
          },
          "ref": null,
          "rendered": "Send",
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
